lighthouse-core/gather/gather-runner.js

```diff
@@ -491,14 +501,21 @@ class GatherRunner {
 
   static instantiateGatherers(passes, rootPath) {
     return passes.map(pass => {
-      pass.gatherers = pass.gatherers.map(gatherer => {
+      pass.gatherers = pass.gatherers.map(gathererDefn => {
         // If this is already instantiated, don't do anything else.
-        if (typeof gatherer !== 'string') {
-          return gatherer;
+        if (gathererDefn.instance) {
+          return gathererDefn;
+        }
+
+        if (typeof gathererDefn.beforePass === 'function') {
+          return {instance: gathererDefn, options: {}};
         }
 
-        const GathererClass = GatherRunner.getGathererClass(gatherer, rootPath);
-        return new GathererClass();
+        const gathererPath = typeof gathererDefn === 'string' ? gathererDefn : gathererDefn.path;
+        const GathererClass = GatherRunner.getGathererClass(
+          gathererDefn.implementation || gathererPath, rootPath);
```

Comment:
should we move all this into config.js? seems like it spiritually belongs over there.

There's also a lot of cases (6+?) being served by these lines: https://github.com/GoogleChrome/lighthouse/blob/c2652e6851a0b9b1a33fab978bfc4cd99bb4db75/lighthouse-core/gather/gather-runner.js#L505-L516

I'm hoping by moving it into config, we're going to end up with something simpler. Open to ideas.

Reply:
yeah just lots of test changes 😢 I'm on it though